test/distributed/test_c10d_error_logger.py

```diff
@@ -1,17 +1,143 @@
+        self.dist_init()
+        func(self)
+        self.destroy_comms()
+
+    return wrapper
+
+
+class C10dErrorLoggerTest(MultiProcessTestCase):
+    def setUp(self):
+        super(C10dErrorLoggerTest, self).setUp()
+        os.environ["WORLD_SIZE"] = str(self.world_size)
+        os.environ["BACKEND"] = BACKEND
+        self._spawn_processes()
+
+    @property
+    def device(self):
+        return (
+            torch.device(self.rank)
+            if BACKEND == dist.Backend.NCCL
+            else torch.device("cpu")
+        )
+
+    @property
+    def world_size(self):
+        return WORLD_SIZE
+
+    @property
+    def process_group(self):
+        return dist.group.WORLD
 
-class C10dErrorLoggerTest(unittest.TestCase):
+    def destroy_comms(self):
+        # Wait for all ranks to reach here before starting shutdown.
+        dist.barrier()
+        dist.destroy_process_group()
 
-    @patch("torch.distributed.c10d_error_logger._get_logging_handler")
-    def test_get_or_create_logger(self, logging_handler_mock):
-        logging_handler_mock.return_value = logging.NullHandler(), "NullHandler"
+    def dist_init(self):
+        dist.init_process_group(
+            backend=BACKEND,
+            world_size=self.world_size,
+            rank=self.rank,
+            init_method=f"file://{self.file_name}",
+        )
+
+        # set device for nccl pg for collectives
+        if BACKEND == "nccl":
+            torch.cuda.set_device(self.rank)
+
+    def test_get_or_create_logger(self):
         logger = _get_or_create_logger()
         self.assertIsNotNone(logger)
         self.assertEqual(1, len(logger.handlers))
         self.assertIsInstance(logger.handlers[0], logging.NullHandler)
+
+    @exception_handler
+    def failed_broadcast_raise_exception(self):
+        tensor = torch.arange(2, dtype=torch.int64)
+        dist.broadcast(tensor, self.world_size + 1)
+
+    @exception_handler
+    def failed_broadcast_not_raise_exception(self):
+        try:
+            tensor = torch.arange(2, dtype=torch.int64)
+            dist.broadcast(tensor, self.world_size + 1)
+        except Exception as exception:
+            pass
+
+    @with_comms
+    def test_exception_handler_with_dist(self) -> None:
+        with self.assertRaises(Exception) as exception:
+            self.failed_broadcast_raise_exception()
+
+        with self.assertLogs(dist._c10d_error_logger, level="DEBUG") as captured:
+            self.failed_broadcast_not_raise_exception()
+            error_msg_dict = json.loads(
+                re.search("({.+})", captured.output[0]).group(0).replace("'", '"')
+            )
+            self.assertEqual(len(error_msg_dict), 7)
+
+            self.assertIn("func_name", error_msg_dict.keys())
+            self.assertEqual("broadcast", error_msg_dict["func_name"])
+
+            self.assertIn("args", error_msg_dict.keys())
+            self.assertEqual("(tensor([0, 1]), 5), {}", error_msg_dict["args"])
```

Comment:
Seeing a failure for this test here: https://github.com/pytorch/pytorch/actions/runs/3325425880/jobs/5498283134. The issue is some CI machines use GPU and have only 2 devices, so the WORLD_SIZE becomes 3 due to your logic on line 31. You can either change this assert to do 
```python
self.assertEqual(f"(tensor([0, 1]), {WORLD_SIZE+1}), {{}}", error_msg_dict["args"])
```
or update WORLD_SIZE to always be 2 for all cases and hardcode the assert.